posthog/hogql/database/schema/util/session_where_clause_extractor.py

```diff
@@ -0,0 +1,329 @@
+            return True
+        # some functions return a constant if the first argument is a constant
+        if node.name in [
+            "parseDateTime64BestEffortOrNull",
+            "toDateTime",
+            "toTimeZone",
+            "assumeNotNull",
+            "toIntervalYear",
+            "toIntervalMonth",
+            "toIntervalWeek",
+            "toIntervalDay",
+            "toIntervalHour",
+            "toIntervalMinute",
+            "toIntervalSecond",
+            "toStartOfDay",
+            "toStartOfWeek",
+            "toStartOfMonth",
+            "toStartOfQuarter",
+            "toStartOfYear",
+        ]:
+            return self.visit(node.args[0])
+
+        if node.name in ["minus", "add"]:
+            return all(self.visit(arg) for arg in node.args)
+
+        # otherwise we don't know, so return False
+        return False
+
+    def visit_field(self, node: ast.Field) -> bool:
+        return False
+
+    def visit_and(self, node: ast.And) -> bool:
+        return False
+
+    def visit_or(self, node: ast.Or) -> bool:
+        return False
+
+    def visit_not(self, node: ast.Not) -> bool:
+        return False
+
+    def visit_placeholder(self, node: ast.Placeholder) -> bool:
+        raise Exception()
+
+    def visit_alias(self, node: ast.Alias) -> bool:
+        return self.visit(node.expr)
+
+
+def is_simple_timestamp_field_expression(expr: ast.Expr) -> bool:
+    return IsSimpleTimestampFieldExpressionVisitor().visit(expr)
+
+
+class IsSimpleTimestampFieldExpressionVisitor(Visitor):
+    def visit_constant(self, node: ast.Constant) -> bool:
+        return False
+
+    def visit_field(self, node: ast.Field) -> bool:
+        # this is quite leaky, as it doesn't handle aliases, but will handle all of posthog's hogql queries
+        return (
+            node.chain == ["min_timestamp"]
+            or node.chain == ["sessions", "min_timestamp"]
+            or node.chain == ["s", "min_timestamp"]
+            or node.chain == ["timestamp"]
+            or node.chain == ["events", "timestamp"]
+            or node.chain == ["e", "timestamp"]
+        )
+
+    def visit_arithmetic_operation(self, node: ast.ArithmeticOperation) -> bool:
+        # only allow the min_timestamp field to be used on one side of the arithmetic operation
+        return (
+            self.visit(node.left)
+            and is_time_or_interval_constant(node.right)
+            or (self.visit(node.right) and is_time_or_interval_constant(node.left))
+        )
+
+    def visit_call(self, node: ast.Call) -> bool:
+        # some functions count as a timestamp field expression if their first argument is
+        if node.name in [
+            "parseDateTime64BestEffortOrNull",
+            "toDateTime",
+            "toTimeZone",
+            "assumeNotNull",
+            "toStartOfDay",
+            "toStartOfWeek",
+            "toStartOfMonth",
+            "toStartOfQuarter",
+            "toStartOfYear",
+        ]:
```

Comment:
This is where our improved type system is much needed.. remembering to keep this list up to date as we add more clickhouse functions will be hard